File: 0207-course-schedule/0207-course-schedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    var indegree = new Array(numCourses).fill(0);
    var list = indegree.map(() => [])
    for (var i = 0; i < prerequisites.length; i++) {
        var pair = prerequisites[i];
        list[pair[1]].push(pair[0]);
        indegree[pair[0]]++;
    }
    var stack = [];
    var count = 0;
    for (var i = 0; i < indegree.length; i++) {
        if (indegree[i] === 0) {
            stack.push(i);
        }
    }
    while (stack.length) {
        var curr = stack.pop();
        count++;
        const adj = list[curr]
        adj.forEach(neighbor => {
            indegree[neighbor]--;
            if (indegree[neighbor] === 0) {
                stack.push(neighbor);
            }
        })
    }
    return count === numCourses;
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegree = {i:0 for i in range(numCourses)}
        adjList = {i: [] for i in range(numCourses)}
        count = 0
        for n1, n2 in prerequisites:
            adjList[n2].append(n1)
            indegree[n1] += 1
        stack = []
        for i in range(len(indegree)):
            if indegree[i] == 0:
                stack.append(i)
        while stack:
            curr = stack.pop()
            count += 1
            adj = adjList[curr]
            for i in range(len(adj)):
                n = adj[i]
                indegree[n] -= 1
                if indegree[n] == 0:
                    stack.append(n)
        return count == numCourses